### PR TITLE
lib/types: Bracket the nested type descriptions for easier reading

### PR DESCRIPTION
**Pull request by Profpatsch**
Type descriptions were constructed in a way that did not make the
nesting clear. Example:

lazy attribute set of lazy attribute set of string or boolean or
signed integer or list of string or boolean or signed integer or lazy
attribute set of string or boolean or signed integer or list of string
or boolean or signed integer

possible readings:

(lazy attribute set of (lazy attribute set of string)) or boolean or
signed integer or list of string or boolean or signed integer or lazy
attribute set of string or boolean or signed integer or list of string
or boolean or signed integer

lazy attribute set of ((lazy attribute set of (string or boolean)) or
signed integer or list of string or boolean or signed integer or lazy
attribute set of string or boolean or signed integer or list of string
or boolean or signed integer)

Just to name a few.

Now we just put every nested type into parens, which should make it
possible to decipher the output.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/types.nix [hunk 1]
```diff
@@ -727,7 +727,7 @@ rec {
     # Either value of type `t1` or `t2`.
     either = t1: t2: mkOptionType rec {
       name = "either";
-      description = "${t1.description} or ${t2.description}";
+      description = "(${t1.description}) or (${t2.description})";
```
Comment (roberth) [inline lib/types.nix:730]: `either` is associative, so we don't need to parenthesize its children when chained, when the child `.name` is `"either"`.

File: lib/types.nix [hunk 2]
```diff
@@ -397,7 +397,7 @@ rec {
 
     listOf = elemType: mkOptionType rec {
       name = "listOf";
-      description = "list of ${elemType.description}";
+      description = "list of (${elemType.description})";
```
Comment (roberth) [inline lib/types.nix:400]: Atomic types don't need parentheses in container-like types.

**Review by roberth — COMMENTED** [review]
Adding some parentheses is a good idea, but adding them everywhere is not necessary for resolving ambiguities and unnecessary noise won't help with understand and may well be counterproductive.

I think we can make a great improvement by adding one attribute to the types: `descriptionIsAtomic`, which is `true` for types that won't lead to ambiguities when put into a larger context, such as the "or" conjunction of `either`.

I think you'll want to do `enum` and `nullOr` as well.

File: lib/types.nix (hunk 1 above)
Comment (Profpatsch) [reply lib/types.nix:730]: That is true, but in practice long eithers are very hard to parse as well (see example)

File: lib/types.nix (hunk 2 above)
Comment (Profpatsch) [reply lib/types.nix:400]: See comment, improving this logic would be nice but needs the types themselves to tell whether they want to be bracketed, I didn’t want to make that nontrivial change.

Comment (piegamesde) [reply lib/types.nix:400]: One could do `"(list of ${elemType.description})"` and similarly for all container types. This would easily give use the correct behavior on primitive types, at the cost of having unnecessary parentheses at the top level (which could probably be stripped).

Comment (roberth) [reply lib/types.nix:400]: Here's a way to do it https://github.com/NixOS/nixpkgs/pull/191353
The tests show a few examples.

**Comment by Profpatsch** [comment]
Closed in favor of https://github.com/NixOS/nixpkgs/pull/191353
